django/contrib/auth/middleware.py

```diff
@@ -118,3 +118,23 @@ def _remove_invalid_user(self, request):
         else:
             if isinstance(stored_backend, RemoteUserBackend):
                 auth.logout(request)
+
+
+class OptionalRemoteUserMiddleware(RemoteUserMiddleware):
+    """
+    Middleware for Web-server-provided authentication on logon pages.
+
+    Like RemoteUserMiddleware but keeps the user authenticated even if
+    the header (``REMOTE_USER``) is not found in the request. Useful
+    for setups when the external authentication via ``REMOTE_USER``
+    is only expected to happen on some "logon" URL and the rest of
+    the application wants to use Django authentication mechanism.
+    """
+
+    def process_request(self, request):
+        if hasattr(request, 'user') and request.user.is_authenticated():
+            stored_backend = load_backend(request.session.get(auth.BACKEND_SESSION_KEY, ''))
+            if isinstance(stored_backend, RemoteUserBackend):
+                request.META[self.header] = request.user.get_username()
```

Comment:
This implementation seems less than ideal. We shouldn't add things to `request.META` that weren't really in the request env; this could be misleading to other middleware or view code.

Seems to me we should instead make the "force logout if no header" behavior in `RemoteUserMiddleware` conditional on a class attr which defaults to `True`, then this subclass wouldn't need to do anything but override that class attr.